src/overmap.cpp

```diff
@@ -3719,27 +3879,100 @@ void overmap::connect_closest_points( const std::vector<point> &points, int z, c
     }
 }
 
-// Polish does both good_roads and good_rivers (and any future polishing) in
+void overmap::rail_build_connection( const point &source, const point &dest, int z, const int_id<oter_type_t> &type_id )
+{
+    const int disp = type_id == oter_type_id( "railroad" ) ? 5 : 2;
+
+    const auto estimate = [&]( const pf::node &prev, const pf::node &cur ) {
+        // Reject nodes that don't allow railroads to cross them (e.g. buildings)
+        const auto &id( ter( cur.x, cur.y, z ) );
+
+        if( !railroad_allowed( id ) ) {
+            return -1;
+        }
+        // Reject nodes that make corners on the river
+        if( prev.dir != cur.dir && ( is_river( ter( prev.x, prev.y, z ) ) ||
+                                     is_river( ter( cur.x, cur.y, z ) ) ) ) {
+            return -1;
+        }
+
+        int res = ( std::abs( dest.x - cur.x ) + std::abs( dest.y - cur.y ) ) / disp;
+        // Prefer existing roads.
+        res += id->type_is( type_id ) ? 0 : 3;
+        // Prefer flat land over bridges
+        res += !is_river( id ) ? 0 : 2;
+        // Try not to turn too much
+        //res += (mn.d == d) ? 0 : 1;
+        return res;
+    };
+
+    const oter_id railbridge_ns( "railbridge_ns" );
+    const oter_id railbridge_ew( "railbridge_ew" );
+
+    for( const auto &node : pf::find_path( source, dest, OMAPX, OMAPY, estimate ) ) {
+        auto &id( ter( node.x, node.y, z ) );
+
+        if( is_river( id ) ) {
+            id = node.dir == 1 || node.dir == 3 ? railbridge_ns : railbridge_ew;
+        } else {
+            // @todo Eliminate discrepancy which requires casting and rotation. That is, make 'node' support 'om_direction'.
+            const om_direction::type dir( om_direction::turn_left( static_cast<om_direction::type>( node.dir ) ) );
+            const size_t prev_line( id->type_is( type_id ) ? id->get_line() : 0 );
+
+            id = type_id->get_linear( om_lines::set_segment( prev_line, dir ) );
+        }
+    }
+}
```

Comment:
If this is a copy of build_road_connection or something similar, it will need to be handled by road jsonization. It must not be a copypasted function that duplicates the functionality.

Reply:
That is the copy from the March when there were no road jsonization yet. :)